ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/results/tracing/FlatTraceGenerator.java

```diff
@@ -0,0 +1,296 @@
+    }
+    final List<Trace> flatTraces = new ArrayList<>();
+    tracesContexts.forEach(context -> flatTraces.add(context.getBuilder().build()));
+    traceCounter.incrementAndGet();
+    return flatTraces.stream();
+  }
+
+  private static void handleCall(
+      final TransactionTrace transactionTrace,
+      final TraceFrame traceFrame,
+      final String lastContractAddress,
+      final AtomicLong cumulativeGasCost,
+      final List<Integer> addressVector,
+      final Deque<FlatTrace.Context> tracesContexts,
+      final int traceFrameIndex) {
+    final Bytes32[] stack = traceFrame.getStack().orElseThrow();
+    final Address contractCallAddress = toAddress(stack[stack.length - 2]);
+    if (addressVector.size() <= traceFrame.getDepth()) {
+      addressVector.add(0);
+    } else {
+      addressVector.set(traceFrame.getDepth(), addressVector.get(traceFrame.getDepth()) + 1);
+    }
+    final List<Integer> traceAddress = Lists.newCopyOnWriteArrayList(addressVector);
+    final FlatTrace.Builder subTraceBuilder =
+        FlatTrace.builder().traceAddress(traceAddress).resultBuilder(Result.builder());
+    // get the next trace frame to set the gas field (gas remaining) of the sub trace
+    final TraceFrame traceFrameAfterCall =
+        transactionTrace.getTraceFrames().get(traceFrameIndex + 1);
+    final Action.Builder subTraceActionBuilder =
+        Action.createCallAction(
+            transactionTrace.getTransaction(),
+            lastContractAddress,
+            contractCallAddress,
+            traceFrame,
+            traceFrameAfterCall.getGasRemaining());
+
+    final long gasCost = cumulativeGasCost.longValue();
+    // retrieve the previous transactionTrace context
+    Optional.ofNullable(tracesContexts.peekLast())
+        .ifPresent(
+            previousContext -> {
+              // increment sub traces counter of previous transactionTrace
+              previousContext.getBuilder().incSubTraces();
+              // set gas cost of previous transactionTrace
+              previousContext
+                  .getBuilder()
+                  .getResultBuilder()
+                  .orElse(Result.builder())
+                  .gasUsed(Gas.of(gasCost).toHexString());
+            });
+    tracesContexts.addLast(
+        new FlatTrace.Context(subTraceBuilder.actionBuilder(subTraceActionBuilder)).subTrace());
+    cumulativeGasCost.set(0);
+  }
+
+  private static void handleReturn(
+      final TransactionTrace transactionTrace,
+      final TraceFrame traceFrame,
+      final Optional<String> smartContractAddress,
+      final Deque<FlatTrace.Context> tracesContexts) {
+    final Deque<FlatTrace.Context> polledContexts = new ArrayDeque<>();
+    FlatTrace.Context ctx;
+    boolean continueToPollContexts = true;
+    // find last non returned transactionTrace
+    while (continueToPollContexts && (ctx = tracesContexts.pollLast()) != null) {
+      polledContexts.addFirst(ctx);
+      if (!ctx.isReturned()) {
```

Comment:
I think a guard block instead of a nested if would read better since the if covers the rest of the while block.

```suggestion
      if (ctx.isReturned()) {
        continue;
      }
```

Then remove the matching brace and un indent the block.

Reply:
Nice suggestion. Done.